### ISSUE TITLE
AHBot does not seem to pick items for sale very randomly

### ISSUE THREAD

**Issue by MrSmite** [TrinityCore/TrinityCore#15346]
__Core:__ https://github.com/TrinityCore/TrinityCore/commit/af57efab3d1d7152d5644fc62e423650319d5a0a
__DB:__ TDB_full_335.58_2015_03_21 + updates

__Bug:__

The AHBot does not appear to be selecting items very randomly when posting auctions for sale.

__Details:__

I have my Alliance AH set with the default CONF, except for the following:

* AuctionHouseBot.Items.Amount.Green = 20000
* AuctionHouseBot.Class.Container = 4

After leaving my server running long enough to have the 20,000 green items (db count), I noticed I only have on green quality container, a [Traveler's Bakckpack](http://www.wowhead.com/item=4500/travelers-backpack). Out of __650__ container entries, I have __240__ *Traveler's Backpack*. The rest of the entries are all of common quality.

Not only are there none of the other [uncommon quality](http://www.wowhead.com/items=1?filter=qu=2) bags, there are also no other qualities posted for sale (eg: [rare](http://www.wowhead.com/items=1?filter=qu=3) or [epic](http://www.wowhead.com/items=1?filter=qu=4)) and there are no [specialty bags](http://www.wowhead.com/items=1?filter=ty=10:3:4:5:2:8:7:6:1:9) of any quality.

**Comment by jackpoz** [comment]
ahbot has way more options and you need to check them all. https://github.com/TrinityCore/TrinityCore/blob/d7c53c30fb45f6da015e126aeb8a4a1901f72e58/src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp#L110 does all the checks of exclusion

**Comment by MrSmite** [comment]
@jackpoz Yeah, I have tried several config variants but I always end up with a ton of one item type and nothing of others. The interesting thing is this doesn't seem to effect armor and weapons. I get a very good spread of greens in those areas.

Thanks for the link to the code, I will try running it through the debugger and post back any relevant observations.

**Comment by jackpoz** [comment]
you might want to check the result of the https://github.com/TrinityCore/TrinityCore/blob/d7c53c30fb45f6da015e126aeb8a4a1901f72e58/src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp#L78 query too.
The buy/sell price being 0 is indeed usually the main cause of items not being listed in the AHBot.

**Comment by MrSmite** [comment]
Hmm, that's interesting. Thanks.

I haven't looked into the new AHBot code much so I wasn't aware of that query. Up until a couple weeks ago, I was stripping the new bot code and putting in the old code because I thought the new one was still being worked on. But I got tired of doing that.

**Comment by MrSmite** [comment]
Ok, so I was just looking at the AHBot Seller and it appears that the logic for how the bot determines what gets sold is a little flawed. I just don't have time to actively debug it to find out where.

I think it has to do with the whole *missedItems* vector because this gets priority during *AddNewAuctions*. The problem is that the functions involved in creating this vector seem to limit its contents to items the bot has already posted.

> Here is an example of a current problem I've experienced:

1. I searched the AH for *Winter Squid* to use in crafting of *Grilled Squid*
2. I got zero results
3. I set both level filters to 35 (req. level of *Winter Squid*) and clicked on *Consumable->Food and Drink* (item class 0, subclass 5)
4. I got 8 pages of results.... but they were the same 5 items listed over and over
5. If you query *item_template* for *class 0, subclass 5* requiring level 35 you get 29 items

So despite there being a possibility of 29 consumable food and drink items, the bot gets stuck on the first few items it posts and continues to post them every time the total is less than the CONF.

> Here's what I think is the problem

It seems the problem is with *SetStat()* and *GetMissedItems()*. These are used to populate the AH based on previous auctions posted by the bot for that session. Since my CONF is set to only post 5 items per cycle, the bot simply refreshes those same 5 items once they are low.

> Conclusion

The bot should be picking random items from all possible items, not a vector of previously posted items. The current method severely limits the "randomness* of the items posted if you have a low *ItemsPerCycle* CONF value.

This value however should simply be a limiter but the current implementation ties it to the actual items posted.

**Comment by splinecl** [comment]
Right now it seems like all items are put into 3 categories:
* Vendor items
* Loot item
* Misc, neither vendor or loot

This means that all items that exclusively come from crafting will be in the Misc category.
https://github.com/TrinityCore/TrinityCore/blob/d7c53c30fb45f6da015e126aeb8a4a1901f72e58/src/server/game/AuctionHouseBot/AuctionHouseBotSeller.cpp#L245

What if a fourth Craftable category was introduced? Then there would be a way to to get craftable items for sale without having to enable Misc or add them all to forceInclude.

This should include Glyphs among other things, without all the other junk in Misc. This should also include other missing items like the specialty bags.

So the question is, is there an good way to find a list of all craftable items?
